### ISSUE TITLE
sign .hlkx file

### ISSUE THREAD

**Issue by bill-dtex** [dotnet/SignService#422]
Hello,

Need help with signing Windows HLK submission.  The HLK will create an .hlkx package file that I can sign using signtool.  But AzureSignTool is returning this error: "The file cannot be signed because it is not a recoginized file type for signing or it is corrupt."

**Comment by clairernovotny** [comment]
Are there example files that we can use? At first glance it looks like an OPC file, so we may already have the code for that but I don't know how to create or validate those files. 

https://docs.microsoft.com/en-us/windows-hardware/test/hlk/user/hlk-signing-with-an-hsm#code-samples

/cc @vcsjones

**Comment by bill-dtex** [comment]
Claire,

I am working to provide a hlkx file produced with a sample driver.  As to validation, I only know of looking at the file properties, Digital Signatures tab.

Thanks, Bill

**Comment by bill-dtex** [comment]
Claire,

Attached is a sample hlkx file for signing.  As to validation, I only know of looking at the file properties, Digital Signatures tab.

Thanks, Bill

**Comment by bill-dtex** [comment]
The signing is successful if the .hlkx file is renamed to .psm1  !!!

-Bill

From: Bill Tang
Sent: Tuesday, February 15, 2022 6:10 PM
To: 'dotnet/SignService' ***@***.***>; dotnet/SignService ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [dotnet/SignService] sign .hlkx file (Issue #422)

Claire,

Attached is a sample hlkx file for signing.  As to validation, I only know of looking at the file properties, Digital Signatures tab.

Thanks, Bill

**Comment by clairernovotny** [comment]
Attachments don't seem to be coming through the email reply mechanism. Can you please post it to this issue?

**Comment by bill-dtex** [comment]
Hello Claire,

My organization policy does not allow posting to public forum.  Is there a way to upload the file to you directly?

Is there a beta release of AzureSignTool that accepts the .hlkx file extension.  I can test this.

The hlkx file indeed looks to be an OPC file.  When signed, the signature be can verified by renaming it to a .zip and looking into \package\services\digital-signature

-Bill

**Comment by type4ranjan** [comment]
@bill-dtex After renaming to psm1 and signing with azuresigntool, were you able to submit the signed hlkx file to microsoft hardware submission portal?

**Comment by bill-dtex** [comment]
No, the signature did not work for submission to the Microsoft Hardware Submission Portal.

A simple verification is to rename it back to .hlkx and use file explorer to look at the file properties, digital signatures tab.

According to this article HLK Signing with an HSM<https://docs.microsoft.com/en-us/windows-hardware/test/hlk/user/hlk-signing-with-an-hsm>, it requires signing via the PackageDigitialSignatureManager APIs.  It will be good if this capability is added to SignService.

In the end, we had to abandon using the Azure Key Vault for the submission.  For now, we are using the old HLK SDK PackageManger APIs with our Authenticode certificate (loaded to the Windows Certificate Store).

-Bill

**Comment by type4ranjan** [comment]
@bill-dtex Is it possible to share the code that you have used to load the keys and certificate from Azure Key Vault to windows certificate store and sign it using HLK SDK PackageManger APIs?

**Comment by bill-dtex** [comment]
We are using a non-EV code signing certificate for the HLK SDK PackageManager API that is already installed into the Windows Certificate Store, not extracting from the EV certificate from Azure Key Vault and loading to the Windows Certificate Store.

Here are code snippets using PackageManage API.  But I suspect this will be legacy and the method from here https://docs.microsoft.com/en-us/windows-hardware/test/hlk/user/hlk-signing-with-an-hsm will be the requirement soon.

using Microsoft.Windows.Kits.Hardware.ObjectModel.Submission;

// Sign the package
PackageManager.Sign(sourcePackage, certificate, outputPackage);


private static X509Certificate2 GetCertificateFromStore(string certificateThumbprint, StoreLocation storeLocation)
{
// get first certificate from MY store that matches the thumbprint
}

**Comment by monrapps** [comment]
This works for submission: https://github.com/monrapps/OpenOpcSignTool

**Comment by Jaxelr** [comment]
hey @clairernovotny / @dtivel any idea what's particularly unique about hlkx given that opc is supported? Somehow internally we had a new ask for trusted signing support of hlkx and we thought of using sign cli but noticed this issue was opened.

cc: @TacoTechSharma
